projects/packages/waf/src/class-waf-runtime.php

```diff
@@ -435,95 +457,93 @@ public function unset_var( $key ) {
 	}
 
 	/**
-	 * Meta.
+	 * A cache of metadata about the incoming request.
 	 *
-	 * @param string $key Key.
-	 * @param string $prop Prop.
+	 * @param string $key The type of metadata to request ('headers', 'remote_addr', etc.).
 	 */
-	public function meta( $key, $prop = false ) {
+	public function meta( $key ) {
 		if ( ! isset( $this->metadata[ $key ] ) ) {
 			$value = null;
 			switch ( $key ) {
 				case 'headers':
-					$value = array();
-					foreach ( $_SERVER as $k => $v ) {
-						$k = strtolower( $k );
-						if ( 'http_' === substr( $k, 0, 5 ) ) {
-							$value[ $this->normalize_header_name( substr( $k, 5 ) ) ] = $v;
-						} elseif ( 'content_type' === $k ) {
-							$value['content-type'] = $v;
-						} elseif ( 'content_length' === $k ) {
-							$value['content-length'] = $v;
-						}
-					}
-					$value['content-type'] = ( ! isset( $value['content-type'] ) || '' === $value['content-type'] )
-						// default Content-Type per RFC 7231 section 3.1.5.5.
-						? 'application/octet-stream'
-						: $value['content-type'];
-					$value['content-length'] = ( isset( $value['content-length'] ) && '' !== $value['content-length'] )
-						? $value['content-length']
-						// if the content-length header is missing, default it to zero.
-						: '0';
+					$value = $this->request->get_headers();
+					break;
+				case 'headers_names':
+					$value = $this->args_names( $this->meta( 'headers' ) );
 					break;
 				case 'remote_addr':
-					$value = '';
-					if ( ! empty( $_SERVER['HTTP_CLIENT_IP'] ) ) {
-						$value = wp_unslash( $_SERVER['HTTP_CLIENT_IP'] );
-					} elseif ( ! empty( $_SERVER['HTTP_X_FORWARDED_FOR'] ) ) {
-						$value = wp_unslash( $_SERVER['HTTP_X_FORWARDED_FOR'] );
-					} elseif ( ! empty( $_SERVER['REMOTE_ADDR'] ) ) {
-						$value = wp_unslash( $_SERVER['REMOTE_ADDR'] );
-					}
+					$value = $this->request->get_remote_addr();
 					break;
 				case 'request_method':
-					$value = empty( $_SERVER['REQUEST_METHOD'] )
-						? 'GET'
-						: wp_unslash( $_SERVER['REQUEST_METHOD'] );
+					$value = $this->request->get_method();
 					break;
 				case 'request_protocol':
-					$value = empty( $_SERVER['SERVER_PROTOCOL'] )
-						? ( empty( $_SERVER['HTTPS'] ) ? 'HTTP' : 'HTTPS' )
```

Comment:
Defaulting to "HTTPS" or "HTTP" here is incorrect - if the request protocol isn't available, better to not have anything.